### PR TITLE
Implement federation fields and types using strawberry

### LINKED ISSUE #3521: Wrong info type passed to resolve_reference classmethod
## Describe the Bug

Take the following example:

```py
class User:
    id: strawberry.ID
    email: str

    @classmethod
    def resolve_reference(cls, id: strawberry.ID, info: strawberry.Info) -> Self:
      print(info.selected_fields)
```

This will result in an error saying the attribute doesn't exist, even though it should. The problem appears to stem from the fact that `GraphQLResolveInfo` is passed through instead in this instance rather than `Info` which provides the attribute.

(This is a federated example.)

## System Information

 - Operating system: MacOS (14.4)
 - Strawberry version (if applicable): 0.232.0

## Additional Context

https://discord.com/channels/689806334337482765/1244954381326090331

### PR DESCRIPTION
**Pull request by patrick91**
- **Add tests for federation schema**
- **Implement federation fields and types using strawberry**

Closes #3521 #3522

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

This release refactors our Federation integration to create types using
Strawberry directly, instead of using low level types from GraphQL-core.

The only user facing change is that now the `info` object passed to the
`resolve_reference` function is the `strawberry.Info` object instead of the one
coming coming from GraphQL-core. This is a **breaking change** for users that
were using the `info` object directly.

If you need to access the original `info` object you can do so by accessing the
`_raw_info` attribute.

```python
import strawberry


@strawberry.federation.type(keys=["upc"])
class Product:
    upc: str

    @classmethod
    def resolve_reference(cls, info: strawberry.Info, upc: str) -> "Product":
        # Access the original info object
        original_info = info._raw_info

        return Product(upc=upc)
```



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to @patrick91 for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```

**Review by parafoxia — APPROVED** [review]
Approving wrt #3521, obvs the rest would need another review lmao. Thanks for this!

File: strawberry/federation/schema.py [hunk 1]
```diff
@@ -354,33 +322,54 @@ def _warn_for_federation_directives(self) -> None:
...
+        while isinstance(type_, StrawberryContainer):
+            type_ = type_.of_type
 
-    entity_type = GraphQLUnionType("_Entity", federation_key_types)  # type: ignore
+        if not hasattr(type_, "__strawberry_definition__"):
+            continue
 
-    def _resolve_type(self: Any, value: Any, info: Any) -> str:
-        return self.__strawberry_definition__.name
+        type_definition = get_object_definition(type_, strict=True)
```
Comment (bellini666) [inline strawberry/federation/schema.py:354]: nit: Github will not allow me to comment on all lines, but you can avoid the `hasattr` by using `get_object_definition` directly with `strict=False`

```python
type_definition = get_object_definition(type_, strict=False)
if type_definition is None:
    continue
```

File: strawberry/federation/schema.py [hunk 2]
```diff
@@ -354,33 +322,54 @@ def _warn_for_federation_directives(self) -> None:
...
+
+        if type_definition.is_object_type and _has_federation_keys(type_definition):
+            entity_types.add(type_)
+
+        for field in type_definition.fields:
+            if field.type and field.type not in seen:
+                seen.add(field.type)
+                stack.append(field.type)
```
Comment (bellini666) [inline strawberry/federation/schema.py:362]: nit: reduce one indentation here (but ignore if you prefer the other way)

```suggestion
            if field.type and field.type in seen:
                continue

            seen.add(field.type)
            stack.append(field.type)
```

**Review by bellini666 — APPROVED** [review]
LGTM 👍🏼 

Left some nits, but ignore if you prefer the other way :)

**Pull request merged**